### PR TITLE
Override the authentication listener to validate FORM_SUBMIT

### LINKED ISSUE #558: useRawRequestData leads to empty data
**Affected version(s)**
4.7.x

**Description**  
When validating a form, it repeatedly occurs for me, that the password `$varValue` is `null`.
All other form data is present and will be fetched correctly.

The `FormPassword` uses `useRawRequestData`.
https://github.com/contao/contao/blob/a0057c0bbfa8b85912ca7800af42208f82dc4fd2/core-bundle/src/Resources/contao/forms/FormPassword.php#L61-L66

Widgets with that attribute will be handled differently: https://github.com/contao/contao/blob/a0057c0bbfa8b85912ca7800af42208f82dc4fd2/core-bundle/src/Resources/contao/library/Contao/Widget.php#L770-L776

Introduced in ebb5b0ddcc1ee4da02488beb3994e5e532a91058.

**How to reproduce**  
I do have two cases where this occurs. One of them:

1. A module fragment with Haste form. When validating, the form the password value is `null`. I was able to fix his issue with `$form->getWidget('password')->useRawRequestData = false;`.

Besides this issue that involves a third-party-extension, the issue can be reproduced by doing the following steps on demo.contao.org:

1. Create a registration module (fields: one field being mandatory, e.g. firstname, the password field).
2. Edit the "Not Authenticated 401" page, remove the redirect.
3. Place the module on the 401 page.
4. Fill out the form, firstname persits, password field gets the error "A password has to be at least 8 characters long!".

Screen Shot 2019-07-06 at 17 47 12

**Proposal**  

- It may occur, that the current request (in the sense of `->getCurrentRequest()`) does not have request data ($_POST) at all. Switching to the master request (in the sense of `->getMasterRequest()`) can work.
- Another approach for fixing this would be to get to know, why the current request does not have post values at all in some cases.
- One approach would be to use `Input::postRaw()` instead.

### PR DESCRIPTION
**Pull request by aschempp**
@richardhj can you see if this fixes https://github.com/contao/contao/issues/558 ?

### CODE REVIEW

**Review by Toflar — CHANGES_REQUESTED** [review]
Concept looks logical to me. Needs tests once @richardhj provided feedback.

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use Symfony\Component\Security\Http\Firewall\UsernamePasswordFormAuthenticationListener;
+
+class ContaoLoginAuthenticationListener extends UsernamePasswordFormAuthenticationListener
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function requiresAuthentication(Request $request)
+    {
+        if (!$request->isMethod('POST') || 0 !== strpos((string) $request->request->get('FORM_SUBMIT'), 'tl_login')) {
```
Comment (leofeyer) [inline core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Why are you using `strpos()` here instead of a simple `===`?

**Review by leofeyer — CHANGES_REQUESTED** [review]
The unit tests are still failing and there should be a test for the `contao.security.authentication_listener` service in the `ContaoCoreExtensionTest` class.

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: the frontend login module uses `tl_login_%ID%` as the form submit value.

Comment (leofeyer) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Oh yes, I recall. 👍 

Personally, I would prefer to write the method like this:

```php
protected function requiresAuthentication(Request $request): bool
{
    if (!$request->isMethod('POST') || !$request->request->has('FORM_SUBMIT')) {
        return false;
    }

    return 0 === strpos($request->request->get('FORM_SUBMIT'), 'tl_login');
}
```

WDYT?

Comment (leofeyer) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Oh, and do we have to account for `tl_loginfoo` as well? The method would return `true` currently. Probably change it to `preg_match('/^tl_login(_\d+)?$/', $request->request->get('FORM_SUBMIT'))`?

Comment (aschempp) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: There's no real case of `tl_loginfoo` anywhere, so I don't think we need that complication.

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\Authentication;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Http\Firewall\UsernamePasswordFormAuthenticationListener;
+
+class ContaoLoginAuthenticationListener extends UsernamePasswordFormAuthenticationListener
```
Comment (leofeyer) [inline core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Why did you prefix the class name with `Contao`? It should be `LoginAuthenticationListener` IMHO.

File: core-bundle/tests/DependencyInjection/Security/ContaoLoginFactoryTest.php [hunk 3]
```diff
@@ -52,14 +50,7 @@ public function testConfiguresTheContainerServices(): void
...
-        $this->assertSame(
-            [
-                'lock_period' => 300,
-                'login_attempts' => 3,
-            ],
-            $arguments[0]
-        );
+        $this->assertSame(['lock_period' => 300, 'login_attempts' => 3], $arguments[0]);
 
         $this->assertTrue($container->hasDefinition('security.authentication.listener.form.contao_frontend'));
```
Comment (leofeyer) [inline core-bundle/tests/DependencyInjection/Security/ContaoLoginFactoryTest.php]: I think this line and the lines below need to be removed now that you have removed the `ContaoLoginFactory::createListener()` method, don't they? Should also fix the unit tests.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php]: Because the firewall listener/factory is called `contao_login`

**Comment by aschempp** [comment]
I have update the basic PR features according to my findings yesterday with @richardhj 

The current frontend login implementation works around several issues Contao core has with Symfony security. Symfony assumes there's a URL where you POST the login data, and that there is a page where the login form is. By default, Symfony will redirect a user to the `login_path`, and only check credentials on the `check_path`. For that case, we created the `_contao/login` route, and by setting `use_forward` the system didn't actually redirect but return a subrequest response of that page.

All of this is not actually how Contao works. In the Contao front end, one can place the login module on any page, and on that page the credentials should be checked. We also do not know a login redirect URL, we merely render the 401/403 page type *in place*.

This is what this PR finally implements now: instead of using the default authentication entry point, which redirects or internally forwards to a `login_path`, we simply generate the 401 content in that place. Instead of only checking the authentication on the `check_path`, we override the authentication listener to check whenever a POST request with FORM_SUBMIT=tl_login is sent. We therefore do not need the `contao_frontend_login` route anymore 🎉 

This PR will collide with #1164 and I will rebase and fix issues and tests once #1164 is merged.

**Review by richardhj — DISMISSED** [review]
Very good! This fixes #558 and #966. Sadly #967 persists but I will evaluate once #1164 and this PR is merged.

**Comment by aschempp** [comment]
This PR is feature-complete now, but I need to fix tests etc. Basically, we're moving away from configuring Symfony from behaving as needed by Contao, and simply implement the necessary 20% ourselves.

To summarize the changes and make them more "readable":

1. I have added a custom authentication listener (configured through the security factory). The listener will check for `$_POST['FORM_SUBMIT'] === 'tl_login'` instead of a check path. This makes `check_path` obsolete.

2. As mentioned before, our frontend login route (`/_contao/login`)  has become obsolete thanks to generating the 401 page directly in our security entry point. This makes `login_path` and `use_forward` as well as the `contao_frontend_login` route obsolete.

3. Thanks to no.2 we do no longer need a `failure_path`, because that's the place the login form is on. That's now always the current URL, and I adjusted and simplified our `AuthenticationFailureHandler` accordingly.

4. By handling base64 directly in the `AuthenticationSuccessHandler`, we do not need the `Base64RedirectController` anymore. As we **know** Contao always uses the `_target_path`, we don't need configurations for `default_target_path`.

5. The authentication success and failure handlers are now configured by the security factory, so the `success_handler` and `failure_handler` are no longer needed.

This will collide with #1130 so @bytehead needs to rebase (mostly because I already removed the lock_period etc. config from the security factory.

**Review by Toflar — COMMENTED** [review]
Oh wow, this makes it so much easier for me to understand what's actually going on in the security authentication process 😄 Looks correct to me!

File: core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php [hunk 4]
```diff
@@ -12,40 +12,30 @@
...
-class ContaoLoginFactory extends FormLoginFactory
+class ContaoLoginFactory extends AbstractFactory
 {
     public function __construct()
     {
-        parent::__construct();
-
-        $this->addOption('lock_period', 300);
-        $this->addOption('login_attempts', 3);
+        $this->options = ['require_previous_session' => false];
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php:24]: This unsets the other two options `check_path` and `use_forward`. Is this intended?

File: core-bundle/src/Resources/config/services.yml [hunk 5]
```diff
@@ -578,10 +572,23 @@ services:
     contao.security.authentication_failure_handler:
         class: Contao\CoreBundle\Security\Authentication\AuthenticationFailureHandler
         arguments:
-            - '@http_kernel'
+            - '@?logger'
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/services.yml:575]: This is marked as optional dependency but it is a mandatory constructor argument in the class.

**Review by leofeyer — CHANGES_REQUESTED** [review]
PR looks very good. 👍 The tests are failing though.

File: core-bundle/src/Resources/config/services.yml (hunk 5 above)
Comment (aschempp) [reply core-bundle/src/Resources/config/services.yml:575]: fixed in a507cf7a1e81c189e37294aa3346b3b2f0b7f77f

**Review by Toflar — DISMISSED** [review]
LGTM

File: core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php:24]: Yes, they will be gone in #1130 anyway. We also unset any other options, as they do not make sense for anything in Contao.

**Review by richardhj — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**